src/librustc_mir/transform/simplify_cfg.rs

```diff
@@ -70,155 +66,155 @@ impl<'l> Pass for SimplifyCfg<'l> {
-
-            // See if we can merge the target block into this one
-            loop {
-                let mut inner_change = false;
-
-                if let TerminatorKind::Goto { target } = terminator.kind {
-                    // Don't bother trying to merge a block into itself
-                    if target == bb {
-                        break;
-                    }
-
-                    let num_insts = mir[target].statements.len();
-                    match mir[target].terminator().kind {
-                        TerminatorKind::Goto { target: new_target } if num_insts == 0 => {
-                            inner_change = true;
-                            terminator.kind = TerminatorKind::Goto { target: new_target };
-                            pred_count[target] -= 1;
-                            pred_count[new_target] += 1;
-                        }
-                        _ if pred_count[target] == 1 => {
-                            inner_change = true;
-                            let mut stmts = Vec::new();
-                            {
-                                let target_data = &mut mir[target];
-                                mem::swap(&mut stmts, &mut target_data.statements);
-                                mem::swap(&mut terminator, target_data.terminator_mut());
-                            }
-
-                            mir[bb].statements.append(&mut stmts);
-                        }
-                        _ => {}
-                    };
-                }
-
-                for target in terminator.successors_mut() {
-                    let new_target = match final_target(mir, *target) {
-                        Some(new_target) => new_target,
-                        None if mir[bb].statements.is_empty() => bb,
-                        None => continue
-                    };
-                    if *target != new_target {
-                        inner_change = true;
-                        pred_count[*target] -= 1;
-                        pred_count[new_target] += 1;
-                        *target = new_target;
-                    }
-                }
-
-                changed |= inner_change;
-                if !inner_change {
-                    break;
-                }
-            }
+pub struct CfgSimplifier<'a, 'tcx: 'a> {
+    basic_blocks: &'a mut IdxVec<BasicBlock, BasicBlockData<'tcx>>,
+    pred_count: IdxVec<BasicBlock, u32>
+}
 
-            mir[bb].terminator = Some(terminator);
+impl<'a, 'tcx: 'a> CfgSimplifier<'a, 'tcx> {
+    fn new(mir: &'a mut Mir<'tcx>) -> Self {
+        let mut pred_count = IdxVec::from_elem(0u32, mir.basic_blocks());
 
-            for succ in mir[bb].terminator().successors().iter() {
-                if seen.insert(succ.index()) {
-                    worklist.push(*succ);
+        // we can't use mir.predecessors() here because that counts
+        // dead blocks, which we don't want to.
```

Comment:
to be clear, we will be adjusting this count to keep it up to date, right?

I guess this code is simple enough it's fine to cut-and-paste. In theory, it might be nice if we could "re-use" some underlying predecessor abstraction independent from the cache.